samples/web/content/apprtc/gcm_register.py

```diff
@@ -0,0 +1,222 @@
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def post(self, cmd):
+    if cmd == 'new':
+      self.handle_new()
+    elif cmd == 'update':
+      self.handle_update()
+    elif cmd == 'verify':
+      self.handle_verify()
+    elif cmd == 'del':
+      self.handle_del()
+    elif cmd == 'query':
+      self.handle_query()
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+class GCMRecord(ndb.Model):
+  """Models a user ID to GCM ID binding record."""
+  user_id = ndb.StringProperty()
+  gcm_id = ndb.StringProperty()
+  code = ndb.StringProperty()
+  code_sent_time = ndb.DateTimeProperty()
+  verified = ndb.BooleanProperty()
+
+  @staticmethod
+  def get_by_user_id(user_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.user_id == user_id).fetch()
+
+  @staticmethod
+  def get_by_gcm_id(gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.gcm_id == gcm_id).fetch()
+
+  @staticmethod
+  def has(user_id, gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    return len(q.fetch()) > 0
+
+  @staticmethod
+  @ndb.transactional(retries=100)
+  def add_or_update(user_id, gcm_id, code):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    records = q.fetch()
+    if len(records) == 0:
+      record = GCMRecord(parent = get_ancestor_key(),
+                         user_id = user_id,
+                         gcm_id = gcm_id,
+                         code = code,
+                         verified = False,
+                         code_sent_time = datetime.datetime.now())
+      record.put()
+      logging.info(
+          'GCM binding added, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+      return constants.RESPONSE_CODE_SENT
+
+    assert len(records) == 1
+    if records[0].verified:
+      logging.warning('Cannot update GCM binding code since already verified, '\
+          'user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+      return constants.RESPONSE_INVALID_STATE
+
+    records[0].code = code
+    records[0].code_sent_time = datetime.datetime.now()
+    records[0].put()
+    logging.info(
+          'GCM binding code updated, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+    return constants.RESPONSE_CODE_RESENT
+
+  @staticmethod
+  @ndb.transactional(retries=100)
+  def verify(user_id, gcm_id, code):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
```

Comment:
Maybe helper function to generate this query since it's used several times + fetched.